nixos/modules/services/web-apps/hledger.nix

```diff
@@ -0,0 +1,137 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+
+  cfg = config.services.hledger;
+
+in
+
+{
+  options = {
+    services.hledger = {
+      web = {
+        enable = mkEnableOption "hledger-web accounting webserver";
+        listenHost = mkOption {
+          type = types.str;
+          default = "127.0.0.1";
+          description = ''
+            Address and port this hledger-web instance listens to.
+          '';
+        };
+        listenPort = mkOption {
+          type = types.str;
+          default = "5000";
+          description = ''
+            Port this hledger-web instance listens to.
+          '';
+        };
+        baseURL = mkOption {
+          type = types.str;
+          default = "http://127.0.0.1:${cfg.web.listenPort}/";
+          description = ''
+            Base URL of this hledger-web instance.
+          '';
+        };
+      };
+      api = {
+        enable = mkEnableOption "hledger-api accounting apiserver";
+        listenHost = mkOption {
+          type = types.str;
+          default = "127.0.0.1";
+          description = ''
+            Address and port this hledger-api instance listens to.
+          '';
+        };
+        listenPort = mkOption {
+          type = types.str;
+          default = "8001";
+          description = ''
+            Port this hledger-api instance listens to.
+          '';
+        };
+      };
+
+      statePath = mkOption {
+        type = types.str;
+        default = "/var/lib/hledger";
+        description = "hledger working directory";
+      };
+
+      stateFileName = mkOption {
+        type = types.str;
+        default = ".hledger.journal";
+        description = "hledger filename";
+      };
+
+      user = mkOption {
+        type = types.str;
+        default = "hledger";
+        description = ''
+          User which runs the hledger-web service.
+        '';
+      };
+
+      group = mkOption {
+        type = types.str;
+        default = "hledger";
+        description = ''
+          Group which runs the hledger-web service.
+        '';
+      };
+
+    };
+  };
+
+  config = mkMerge [ (mkIf (cfg.web.enable || cfg.api.enable) {
+    users.extraUsers = optionalAttrs (cfg.user == "hledger") (singleton {
+      name = "hledger";
+      group = cfg.group;
+      uid = config.ids.uids.hledger;
+      home = cfg.statePath;
+      createHome = true;
+    });
+
+    users.extraGroups = optionalAttrs (cfg.group == "hledger") (singleton {
+    name = "hledger";
+    gid = config.ids.gids.hledger;
+    });
+
+  }) (mkIf cfg.web.enable {
+    systemd.services.hledger-web = {
+      description = "hledger-web accounting";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      preStart = ''
+      touch ${cfg.statePath}/${cfg.stateFileName}
```

Comment:
If `/var/lib/hledger` does not exist, this line will fail.
This requires `mkdir -p ${cfg.statePath}`, followed by `touch` and `chown` of the `stateFileName`.

Reply:
Better to use ```StateDirectory = "hledger";``` which will handle the creation and chowing automatically.

Reply:
In general I would agree, however for this service it seems that somebody might prefer to read the file from their own home.